pkg/controllers/allocation/scheduling/scheduler.go

```diff
@@ -101,18 +100,18 @@ func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha4.Provisioner
 }
 
 func (s *Scheduler) solve(ctx context.Context, constraints *v1alpha4.Constraints, pods []*v1.Pod) ([]*Schedule, error) {
-	// Apply runtime constraints
+	// Consolidate requirements in memory before executing scheduling logic.
+	// This is a performance optimization to avoid executing requirement
+	// evaluation logic redundantly for every pod.
 	constraints = constraints.DeepCopy()
-	if err := constraints.Constrain(ctx); err != nil {
-		return nil, fmt.Errorf("applying constraints, %w", err)
-	}
+	constraints.Requirements = constraints.CombineRequirements()
```

Comment:
Seems a bit odd for the constraints struct not to handle this itself (as opposed to the Scheduler reaching in and doing it?)

Reply:
In the first example, we would need to pass these combined requirements (along w/ provider, taints, and labels) into the per-pod scheduling logic. By the time we're passing those 4 things, we might as well use the constraints object, since it has those exact fields.

I like the second approach you noted.